test/integration/jobset_controller_test.go

```diff
@@ -311,6 +329,77 @@ var _ = ginkgo.Describe("JobSet controller", func() {
 			gomega.Eventually(checkJobSetStatus, timeout, interval).WithArguments(js, jobset.JobSetFailed).Should(gomega.Equal(true))
 		})
 	})
+
+	ginkgo.When("a jobset is created with failure policy 'Any' and restart policy 'RecreateAll'", func() {
+		ginkgo.It("should create all jobs, then restart every job once any job fails", func() {
+			ginkgo.By("creating a new JobSet")
+			ctx := context.Background()
+
+			// Construct JobSet with 2 replicated jobs with 3 replicas each.
+			js := testing.MakeJobSet("js-failure-policy-any-with-recreate", ns.Name).
+				AddReplicatedJob(testing.MakeReplicatedJob("replicated-job-leader").
+					SetJob(testing.MakeJob("test-job-leader", ns.Name).
+						SetCompletionMode(batchv1.IndexedCompletion).Obj()).
+					SetReplicas(3).
+					Obj()).
+				AddReplicatedJob(testing.MakeReplicatedJob("replicated-job-worker").
+					SetJob(testing.MakeJob("test-job-worker", ns.Name).
+						SetCompletionMode(batchv1.IndexedCompletion).Obj()).
+					SetReplicas(3).
+					Obj()).
+				// Set failure policy to "Any" with restart policy "Recreate" with max 1 restart.
+				SetFailurePolicy(&jobset.FailurePolicy{
+					Operator:      jobset.TerminationPolicyTargetAny,
+					RestartPolicy: jobset.RestartPolicyRecreateAll,
+					MaxRestarts:   1,
+				}).
+				Obj()
+
+			// Create JobSet.
+			gomega.Expect(k8sClient.Create(ctx, js)).Should(gomega.Succeed())
+
+			// We'll need to retry getting this newly created JobSet, given that creation may not immediately happen.
+			ginkgo.By("checking JobSet was created successfully")
+			gomega.Eventually(k8sClient.Get(ctx, types.NamespacedName{Name: js.Name, Namespace: js.Namespace}, &jobset.JobSet{}), timeout, interval).Should(gomega.Succeed())
+
+			ginkgo.By("checking JobSet eventually has 6 active jobs")
+			var originalJobList batchv1.JobList
+			gomega.Eventually(func() (int, error) {
+				if err := k8sClient.List(ctx, &originalJobList, client.InNamespace(js.Namespace)); err != nil {
+					return -1, err
+				}
+				return len(originalJobList.Items), nil
+			}, timeout, interval).Should(gomega.Equal(6))
+
+			// Make a job fail.
+			ginkgo.By("making 1 job fail")
+			job := originalJobList.Items[0]
+			job.Status.Conditions = append(job.Status.Conditions, batchv1.JobCondition{
+				Type:   batchv1.JobFailed,
+				Status: corev1.ConditionTrue,
+			})
+			gomega.Expect(k8sClient.Status().Update(ctx, &job)).Should(gomega.Succeed())
+
+			// Verify jobs were recreated successfully.
```

Comment:
we need to fail again to reach max restarts and check that the jobset is failed, and on another instance of the test we succeed and verify that the jobset is succeeded, you can add the latter on a followup PR using table format.